### PR TITLE
feat: ADR for groups implementation

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
Description
ADR for the groups implementation into LC

https://2u-internal.atlassian.net/browse/ENT-8336

### CODE REVIEW

File: docs/decisions/0014-access-policy-grouping.rst [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+*******
+The enterprise-access service needs to allow for the flexibility to control and distribute subsidy access policies to
+custom subsets of learners. The subdivision of access policies will improve user management and budgeting from an admin
+perspective, as well as add support for personalization to the learner experience down the line. While it is not
+the responsibility of the service to house the grouping of learners, an access policy must take into account the 
+existence of related subsets when determining if a subsidy is redeemable by an individual learner.
+
+Decision
+********
+The ``SubsidyAccessPolicy`` model will adopt a new, nullable UUID field: `enterprise_group_uuid`. This field will be
```
Comment (kiram15) [inline docs/decisions/0014-access-policy-grouping.rst:18]: Do we want this to be a nullable field? With this change, by default, every single policy will have a group, and all the learners will be under this group.

File: docs/decisions/0014-access-policy-grouping.rst [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+How the systems will interact with one another
+++++++++++++++++++++++++++++++++++++++++++++++
+Upon provisioning a new access policy budget for a customer, the service will make a POST request to edx-platform to
+create a new ``EnterpriseGroup`` record. On successful response, the enterprise-access service will write the returned
+UUID of the newly created group to the access policy `enterprise_group_uuid` field.
+
+``SubsidyAccessPolicy``'s `can_redeem()` method already makes a request to edx-platform for
+`enterprise_contains_learner()` in which `lms_user_id`` and `enterprise_customer_uuid` are provided to confirm
+a learner's membership with the associated organization. Now, `enterprise_group_uuid` will be optionally supplied by
```
Comment (kiram15) [inline docs/decisions/0014-access-policy-grouping.rst:29]: Again, wouldn't this not be optional? We would have to know if the user's group permissions have access to this course

File: docs/decisions/0014-access-policy-grouping.rst (hunk 1 above)
Comment (alex-sheehan-edx) [reply docs/decisions/0014-access-policy-grouping.rst:18]: my thought was backwards compatibility- this way we can have a chance to at a minimum backfill all the existing policies with new groups

Comment (alex-sheehan-edx) [reply docs/decisions/0014-access-policy-grouping.rst:18]: + there's some order of operations stuff going on, ie we need to create the policy then the group then update the policy with the group uuid

File: docs/decisions/0014-access-policy-grouping.rst (hunk 2 above)
Comment (alex-sheehan-edx) [reply docs/decisions/0014-access-policy-grouping.rst:29]: optional in that I think it should be backwards compatible

Comment (alex-sheehan-edx) [reply docs/decisions/0014-access-policy-grouping.rst:29]: so if a policy doesn't have a group it won't blow up

File: docs/decisions/0014-access-policy-grouping.rst (hunk 1 above)
Comment (kiram15) [reply docs/decisions/0014-access-policy-grouping.rst:18]: Yeah that makes sense, but I also think we should call out that it really shouldn't be null in practice

Comment (alex-sheehan-edx) [reply docs/decisions/0014-access-policy-grouping.rst:18]: Sure, I made an addition

Comment (johnnagro) [reply docs/decisions/0014-access-policy-grouping.rst:18]: yeah i think for backwards compatibility we'll want this to be nullable. We can also discuss different ways to implement the "all users in the enterprise" group which is an implicit group today. One approach is the "blank" or "no group specified" means "all users in the enterprise". Alternatively you can have a magic/dynamic group which is all users (if you feel the need to be explicit with a UUID for "all users" of a given enterprise customer).

**Review by johnnagro — APPROVED** [review]

File: docs/decisions/0014-access-policy-grouping.rst (hunk 1 above)
Comment (alex-sheehan-edx) [reply docs/decisions/0014-access-policy-grouping.rst:18]: Agreed and I feel that leaving the field nullable keeps us open to either implementation

**Pull request merged**